### PR TITLE
Snapshots for arbitrary devices and sample-based measurements

### PR DESCRIPTION
**Pull request by Shiro-Raven**
**Context:**
`qml.Snapshot` and `qml.snapshots` are debugging tools that allow users to carry out state measurements at an arbitrary point of the execution. In #4908, support for arbitrary measurements was added, but limited to analytical calculations. In addition, support was limited to only some simulators. Therefore, `qml.snapshots` has to be changed to a transform that is able to handle any device. 

**Description of the Change:**
- For unsupported devices, `qml.snapshots` now splits the `QuantumTape` into several, creating one for each snapshot measurement, aggregating all the operations occurring before that snapshot. The result from each execution is collected into a dictionary as what used to happen before.
- `validate_measurements` is now complemented to take into account the measurements in the snapshots as part of the validations process.
- Support is also added for sample-based measurements which inherit the shots from the device.
- Several bugs have been fixed:
  - The measurement is now passed to the new snapshot instance during adjoint and control operations.
  - An error is now raised if a tag of type other than string to passed, as declared in the docs.
  - For some devices (e.g. `qutrit.mixed`), the state was returned by the snapshot regardless of the requested measurement.
  - For other devices (e.g. `qubit.legacy`), an error is raised when a measurement other than `qml.state` is requested, as modifying the code to be more generic required a lot of effort.
  - For legacy devices, the snapshot handling logic did not reset the `_debugger` attribute after the execution. This is now fixed.
  - `default.qutrit` was falsely flagged as snapshot-compatible through the `_debugger` attribute. This is now commented out.

**Benefits:**
- More consistent snapshot behaviour across devices and clearer error handling for invalid cases.
- Support for QPUs and majorly non-PL/exotic simulators. 

[[sc-50047](https://app.shortcut.com/xanaduai/story/50047)] Supersedes #5722

### CODE REVIEW

File: doc/releases/changelog-dev.md
Comment (trbromley) [inline doc/releases/changelog-dev.md:301]:
```suggestion
* `qml.snapshots` raises an error when a measurement other than `qml.state` is requested from `default.qubit.legacy` instead of silently returning the statevector.
  [(#5805)](https://github.com/PennyLaneAI/pennylane/pull/5805)
```

File: pennylane/debugging.py
Comment (Shiro-Raven) [reply pennylane/debugging.py:134]: yes! my bad

File: pennylane/debugging.py
Comment (Shiro-Raven) [reply pennylane/debugging.py:62]: Of course we can simply just disable pylint for this line

Comment (Jaybsoni) [reply pennylane/debugging.py:62]: The tape function isn't accessible top level, it's just scoped to this file. Can you share the error you were seeing with pylint? Typically disabling pylint if the error is expected and unavoidable is common. Additionally, we could rename the argument aswell. Something like `qtape` or `qscript` if its internal.

Comment (Jaybsoni) [reply pennylane/debugging.py:62]: The decision to rename the `qml.debugging.tape()` function would be up to @DSGuala and @trbromley.

Comment (Shiro-Raven) [reply pennylane/debugging.py:62]: @Jaybsoni pardon my incorrect terminology 😅  The error I got from `pylint` was
 `pennylane/debugging.py:63:14: W0621: Redefining name 'tape' from outer scope (line 416) (redefined-outer-name)`

Comment (mudit2812) [reply pennylane/debugging.py:62]: You could also use `circuit` instead of `tape_`

File: pennylane/debugging.py
Comment (mudit2812) [inline pennylane/debugging.py:63]:
```suggestion
    r"""This transform processes :class:`~pennylane.Snapshot` instances contained in a circuit,
```

File: pennylane/debugging.py
Comment (Shiro-Raven) [reply pennylane/debugging.py:62]: @mudit2812 wouldn't this be a bit inconsistent with the transforms semantic?

Comment (Jaybsoni) [reply pennylane/debugging.py:62]: Yes, I think I am happy with either solution (renaming the variable or disabling pylint for that line). I haven't seen this error appear before, but I'd like to think this must have happened before. Could be worth searching the code base to see how it was handled.

Comment (mudit2812) [reply pennylane/debugging.py:62]: > @mudit2812 wouldn't this be a bit inconsistent with the transforms semantic?

Yes, but so is `tape_`

Comment (albi3ro) [reply pennylane/debugging.py:62]: Every other transform uses `tape` as the first keyword argument, so we should continue calling it `tape` for this transform. Otherwise we would be breaking the `transform` convention.

Can we split the snapshots stuff and the debugger stuff into two separate files?

Comment (Shiro-Raven) [reply pennylane/debugging.py]: @trbromley `breakpoint` and `snapshots` now live in two separate files #5845. I renamed the argument name to `tape`

File: tests/test_debugging.py
Comment (albi3ro) [inline tests/test_debugging.py:58]: I'd prefer to add in a few more explicit tests for the actual content of the batches. 

```suggestion
        assert len(tapes_no_meas) == num_snapshots
        tape1 = qml.tape.QuantumScript([], [qml.state()])
        assert qml.equal(tapes_no_meas[0], tape1)
        tape2 = qml.tape.QuantumScript([qml.Hadamard(0)], [qml.state()])
        assert qml.equal(tapes_no_meas[1], tape2)
        tape3 = qml.tape.QuantumScript([qml.Hadamad(0), qml.CNOT((0,1))], [qml.state()])
        assert qml.equal(tapes_no_meas[2], tape3)
        tape4 = qml.tape.QuantumScript([qml.Hadamard(0), qml.CNOT((0,1))], [qml.expval(qml.X(0))])
        assert qml.equal(tapes_no_meas[3], tape4)
```

I would also recommend having some of the snapshots here include measurements.

A lot of things could be happening inside `qml.snapshot.transform` that still give rise to the correct length of batch.

**Review by albi3ro — COMMENTED** [review]
Thanks for all your patience 🎉 

I think I'd like to see a bit more specificity in the unit tests, but after that I will be happy to approve :)

**Review by albi3ro — APPROVED** [review]
🎉

**Pull request merged**
